pkg/encryption/store/decrypt_store.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package store
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/encryption"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"golang.org/x/crypto/sha3"
+)
+
+type decryptingStore struct {
+	storage.Getter
+}
+
+func NewDecrypting(s storage.Getter) storage.Getter {
+	return &decryptingStore{s}
+}
+
+func (s *decryptingStore) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Address) (ch swarm.Chunk, err error) {
+	switch l := len(addr.Bytes()); l {
+	case swarm.HashSize:
+		// normal, unencrypted content
+		return s.Getter.Get(ctx, mode, addr)
+
+	case encryption.ReferenceSize:
+		// encrypted reference
+		ref := addr.Bytes()
+		address := swarm.NewAddress(ref[:swarm.HashSize])
+		ch, err := s.Getter.Get(ctx, mode, address)
+		if err != nil {
+			return nil, err
+		}
+
+		d, err := decryptChunkData(ch.Data(), ref[swarm.HashSize:])
+		if err != nil {
+			return nil, err
+		}
+		return swarm.NewChunk(address, d), nil
+
+	default:
+		return nil, storage.ErrReferenceLength
+	}
+}
+
+func decryptChunkData(chunkData []byte, encryptionKey encryption.Key) ([]byte, error) {
+	if len(chunkData) < swarm.SpanSize {
+		return nil, fmt.Errorf("invalid ChunkData, min length 8 got %v", len(chunkData))
+	}
+
+	decryptedSpan, decryptedData, err := decrypt(chunkData, encryptionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// removing extra bytes which were just added for padding
+	length := binary.LittleEndian.Uint64(decryptedSpan)
+	refSize := int64(swarm.HashSize + encryption.KeyLength)
+	for length > swarm.ChunkSize {
+		length = length + (swarm.ChunkSize - 1)
+		length = length / swarm.ChunkSize
+		length *= uint64(refSize)
+	}
+
+	c := make([]byte, length+8)
+	copy(c[:8], decryptedSpan)
+	copy(c[8:], decryptedData[:length])
+
+	return c, nil
+}
+
+func decrypt(chunkData []byte, key encryption.Key) ([]byte, []byte, error) {
+	encryptedSpan, err := newSpanEncryption(key).Encrypt(chunkData[:swarm.SpanSize])
+	if err != nil {
+		return nil, nil, err
+	}
+	encryptedData, err := newDataEncryption(key).Encrypt(chunkData[swarm.SpanSize:])
```

Comment:
should this not be `Decrypt` instead of Encrypt? lol  because of symmetry  it works but super confusing

Reply:
var should be just `data`?